ocaml/xapi/message_forwarding.ml

```diff
@@ -6432,4 +6440,139 @@ functor
+        let self =
+          Local.Observer.create ~__context ~name_label ~name_description ~hosts
+            ~attributes ~endpoints ~components ~enabled
+        in
+        let local_fn =
+          Local.Observer.register ~self ~host:(Helpers.get_localhost ~__context)
+        in
+        Xapi_observer.observed_hosts_of ~__context hosts
+        |> List.iter (fun host ->
+               do_op_on ~__context ~host ~local_fn (fun session_id rpc ->
+                   Client.Observer.register ~rpc ~session_id ~self ~host
+               )
+           ) ;
+        self
+
+      let register ~__context ~self ~host =
+        info "Observer.register: self=%s" (observer_uuid ~__context self) ;
+        let local_fn = Local.Observer.register ~self ~host in
+        let client_fn session_id rpc =
+          Client.Observer.register ~rpc ~session_id ~self ~host
+        in
+        do_op_on ~__context ~host ~local_fn client_fn
+
+      let unregister ~__context ~self ~host =
+        info "Observer.unregister: self=%s" (observer_uuid ~__context self) ;
+        let local_fn = Local.Observer.unregister ~self ~host in
+        let client_fn session_id rpc =
+          Client.Observer.unregister ~rpc ~session_id ~self ~host
+        in
+        do_op_on ~__context ~host ~local_fn client_fn
+
+      let destroy ~__context ~self =
+        info "Observer.destroy: self=%s" (observer_uuid ~__context self) ;
+        let hosts = Db.Observer.get_hosts ~__context ~self in
+        let local_fn =
+          Local.Observer.register ~self ~host:(Helpers.get_localhost ~__context)
+        in
+        Xapi_observer.observed_hosts_of ~__context hosts
+        |> List.iter (fun host ->
+               do_op_on ~__context ~host ~local_fn (fun session_id rpc ->
+                   Client.Observer.unregister ~rpc ~session_id ~self ~host
+               )
+           ) ;
+        Local.Observer.destroy ~__context ~self
+
+      let set_hosts ~__context ~self ~value =
+        let uuid = observer_uuid ~__context self in
+        info "Observer.set_hosts: self=%s value=%s" uuid
+          (String.concat ";" (List.map (host_uuid ~__context) value)) ;
+        Local.Observer.set_hosts ~__context ~self ~value ;
+        let new_hosts =
+          RefSet.of_list (Xapi_observer.observed_hosts_of ~__context value)
+        in
+        let old_hosts =
+          RefSet.of_list
+            (Xapi_observer.observed_hosts_of ~__context
+               (Db.Observer.get_hosts ~__context ~self)
+            )
+        in
+        let to_add = RefSet.diff new_hosts old_hosts in
+        let to_remove = RefSet.diff old_hosts new_hosts in
+        RefSet.iter
+          (fun host ->
+            let local_fn = Local.Observer.unregister ~self ~host in
+            do_op_on ~__context ~host ~local_fn (fun session_id rpc ->
+                Client.Observer.unregister ~rpc ~session_id ~self ~host
+            )
+          )
+          to_remove ;
```

Comment:
`local_fn` must be defined outside of the iter

Reply:
I thought I had updated that

Reply:
[600baec](https://github.com/xapi-project/xen-api/pull/4966/commits/600baec4d018797587fb3a25856f7e49429a1f81)